In [ ]:
library(tidyverse)
options(repr.plot.width=5, repr.plot.height=4)

Warning message in system("timedatectl", intern = TRUE):
“running command 'timedatectl' had status 1”
── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──

✔ ggplot2 3.4.0      ✔ purrr   1.0.1 
✔ tibble  3.1.8      ✔ dplyr   1.0.10
✔ tidyr   1.2.1      ✔ stringr 1.4.1 
✔ readr   2.1.3      ✔ forcats 0.5.2 

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



# STATS 306
## Homework 6: Importing data

For each problem, enter the R code in the cell marked "YOUR SOLUTION HERE".

## Problem 1: Scraping Wikipedia
Use Wikipedia and the `rvest` package to answer the following the questions:

In [4]:
library(rvest)

**(a)** What is the distribution of the age of U.S. presidents when they first took office? Produce an appropriate visualization. 

(Note: in cases where you only know the year of birth, assume the president was born on Jan 1 of that year. If a president took office more than once, use the earliest date.)

**(b)** Make a map showing the locations of the ten largest soccer stadiums in the world. 

Note: the following code will load a database of the locations large cities:

In [11]:
library(maps)
data(world.cities)
head(world.cities)

name               country.etc pop   lat   long  capital
1 'Abasan al-Jadidah Palestine    5629 31.31 34.34 0      
2 'Abasan al-Kabirah Palestine   18999 31.32 34.35 0      
3 'Abdul Hakim       Pakistan    47788 30.55 72.11 0      
4 'Abdullah-as-Salam Kuwait      21817 29.36 47.98 0      
5 'Abud              Palestine    2456 32.03 35.07 0      
6 'Abwein            Palestine    3434 32.03 35.20 0

**(c)** Make a plot that compares the total number of TV viewers over time for a) the Super Bowl, b) the World Series, and c) the NBA Finals. (Use the average number of viewers per game for baseball and basketball.)

**(d)** Make a visualization of the 50 highest-grossing films of all time, how much money they made, and what year they were released. (There are multiple ways to accomplish this; you may have to experiment a bit.)

## IMDB 250
The following CSV file contains data on the [top 250 rated movies on IMDB](https://www.imdb.com/chart/top):

In [37]:
imdb250.raw %>% arrange(movie_year)

rank movie_name                                movie_year
1   129  The Kid                                   1921      
2   194  Sherlock Jr.                              1924      
3   175  The Gold Rush                             1925      
4   182  The General                               1926      
5   114  Metropolis                                1927      
6   207  La passion de Jeanne d'Arc                1928      
7    52  City Lights                               1931      
8    97  M - Eine Stadt sucht einen Mörder         1931      
9   241  It Happened One Night                     1934      
10   46  Modern Times                              1936      
11  158  Gone with the Wind                        1939      
12  191  Mr. Smith Goes to Washington              1939      
13  222  The Wizard of Oz                          1939      
14   63  The Great Dictator                        1940      
15  230  The Grapes of Wrath                       1940      
16  231  Rebecca                                   1940      
17   94  Citizen Kane                              1941      
18   42  Casablanca                                1942      
19  229  To Be or Not to Be                        1942      
20  104  Double Indemnity                          1944      
21   21  It's a Wonderful Life                     1946      
22  227  The Best Years of Our Lives               1946      
23  121  Ladri Di Biciclette                       1948      
24  146  The Treasure of the Sierra Madre          1948      
25  179  The Third Man                             1949      
26   60  Sunset Blvd.                              1950      
27  131  All About Eve                             1950      
28  148  Rashōmon                                  1950      
29   86  Singin' in the Rain                       1952      
30  107  Ikiru                                     1952      
⋮   ⋮    ⋮                                         ⋮         
221 186  Gone Girl                                 2014      
222 190  The Grand Budapest Hotel                  2014      
223 200  Relatos Salvajes                          2014      
224 161  Inside Out                                2015      
225 199  Room                                      2015      
226 202  Mad Max: Fury Road                        2015      
227 217  Spotlight                                 2015      
228  85  Kimi No Na Wa.                            2016      
229 127  Dangal                                    2016      
230 197  Hacksaw Ridge                             2016      
231 245  Ah-ga-ssi                                 2016      
232  77  Coco                                      2017      
233 167  Three Billboards Outside Ebbing, Missouri 2017      
234 221  Logan                                     2017      
235  65  Avengers: Infinity War                    2018      
236  69  Spider-Man: Into the Spider-Verse         2018      
237  93  Capharnaüm                                2018      
238 133  Green Book                                2018      
239  35  Parasite                                  2019      
240  72  Joker                                     2019      
241  80  Avengers: Endgame                         2019      
242 122  1917                                      2019      
243 196  Klaus                                     2019      
244 214  Ford v. Ferrari                           2019      
245 119  Hamilton                                  2020      
246 132  The Father                                2020      
247 128  Spider-Man: No Way Home                   2021      
248 248  Jai Bhim                                  2021      
249  45  Top Gun: Maverick                         2022      
250 153  Everything Everywhere All at Once         2022      
    genre                              director_name        rating
1   ["Comedy","Drama","Family"]        Charles Chaplin      8.3   
2   ["Action","Comedy","Ro

In [38]:
# download.file("https://datasets.stats306.org/imdb250.csv", "imdb250.csv")
read_csv('imdb250.csv') %>% print

Rows: 250 Columns: 7
── Column specification ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
Delimiter: ","
chr (5): movie_name, movie_year, genre, director_name, actors_list
dbl (2): rank, rating

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


# A tibble: 250 × 7
    rank movie_name                         movie…¹ genre direc…² rating actor…³
   <dbl> <chr>                              <chr>   <chr> <chr>    <dbl> <chr>  
 1     1 The Shawshank Redemption           1994    "[\"… Frank …   99.3 "[\"Ti…
 2     2 The Godfather                      1972    "[\"… Franci…    9.2 "[\"Ma…
 3     3 The Dark Knight                    2008    "[\"… Christ…    9   "[\"Ch…
 4     4 The Godfather: Part II             1974    "[\"… Franci…    9   "[\"Al…
 5     5 12 Angry Men                       1957    "[\"… Sidney…    9   "[\"He…
 6     6 Schindler's List                   1993    "[\"… Steven…    9   "[\"Li…
 7     7 The Lord of the Rings: The Return… 2003    "[\"… Peter …    9   "[\"El…
 8    16 The Matrix                         1999    "[\"… Lana W…    8.7 "[\"Ke…
 9    17 Goodfellas                         1990    "[\"… Martin…    8.7 "[\"Ro…
10    18 One Flew Over the Cuckoo's Nest    1975    "[\"… Milos …    8.7 "[\"Ja…
# … with

(The CSV contains some data entry errors that you may need to correct.)

**(a)** Make a table or other  visualization that shows the year of release, title, and rating for each movie that Tom Hanks starred in.

**(b)** Make a plot showing the number of top 250 films from each decade. Would you say there was a "golden age" of cinema, during which time comparatively more classic films were released?

**(c)** How many of the top earning films from Problem 1(d) are in the IMDB top 250? Is there a strong connection between performance at the box office and being rated as an all-time great film?



**(d)** The earliest movie in this list is 1921's "The Kid" (rank 129) starring Charlie Chaplin. We could say that "The Kid" was the "reigning champion" on this list for six years, until 1927, when another movie ("Metropolis", rank 114) came along that is ranked higher. "Metropolis" reigned for four years, until another Chaplin Film, "City Lights" (rank 52) came along and dethroned it in 1931. Make a visualization showing the each year and the reigning champion for that year, from 1921 to the present.


## Problem 3 (Challenge problem) 
The following spreadsheet contains data from the [Current Population Survey](https://www.census.gov/data/tables/time-series/demo/educational-attainment/cps-historical-time-series.html) on wages by sex and education level in the U.S.:

https://docs.google.com/spreadsheets/d/1N3-Ez5wDVhUgbjy2zsLULrbrYkpVYvDrf-a88-5ED8Q/edit?usp=sharing

You can download the spreadsheet to your local/Colab folder by executing:

In [41]:
download.file("https://datasets.stats306.org/taba-3.xlsx", 'taba-3.xlsx')
taba3 <- readxl::read_xlsx("taba-3.xlsx")

New names:
• `` -> `...2`
• `` -> `...3`
• `` -> `...4`
• `` -> `...5`
• `` -> `...6`
• `` -> `...7`
• `` -> `...8`
• `` -> `...9`
• `` -> `...10`
• `` -> `...11`
• `` -> `...12`
• `` -> `...13`
• `` -> `...14`
• `` -> `...15`
• `` -> `...16`
• `` -> `...17`
• `` -> `...18`
• `` -> `...19`


As you will see, the data are in a highly non-tidy format. It's your job to tidy them up in order to answer the following questions:

**(a)** Across all races and genders, which demographic group earns the most money? The least? How has this changed over time?

**(b)** Produce an appropriate visualization showing how the wage gap between college-educated men and women has changed from 1975 to the present. Have there been any sudden shifts in the size of the wage gap that you can see?

**(c)** Are women with less education more or less likely to experience a large wage gap compared to highly educated women? Support your reasoning with appropriate statical or visual analysis.

**(d)** In your opinion, will there ever be parity between male and female pay? If so, when? Support your reasoning with statistical evidence.

_Note_: Tidying this dataset is challenging. It's okay to use any strategy that works. For example, you might have to "hardcode" some column names, rather than having R infer them automatically for you. Use whatever techniques you need to answer the questions above, while minimizing the amount of manual data entry you have to perform.